File: dbStuff.py
import sqlite3
from pathlib import Path

# create schema, add, and then read some data
# see/run test.sh for a more robust test that that includes delete cascades

con = sqlite3.connect("test.sqlite3")
con.row_factory = sqlite3.Row  # to return rows as dictionaries
cur = con.cursor()

# ensure blank slate
cur.executescript(Path('sql/test_drop.sql').read_text())

# create tables
cur.execute(Path('sql/user.sql').read_text())
cur.execute(Path('sql/post.sql').read_text())
cur.execute(Path('sql/comment.sql').read_text())

# add some data
cur.executescript(Path('sql/test_insert.sql').read_text())

# read the data back out
tables = ['user', 'post', 'comment']
for table in tables:
    print(f"Records in {table}:")
    cur.execute(f"SELECT * FROM {table}")
    for row in cur.fetchall():
        print(dict(row)) # we still have to cast it =\

con.close()

File: dbStuff.js
const sqlite3 = require('sqlite3').verbose();
let db = new sqlite3.Database('test.sqlite');
var fs = require('fs');

// wrap the db.exec and db.all funcitons so we can have sync execution of our sql

function exec(sql) {
  return new Promise((resolve, reject) => {
      db.exec(sql, (err) => err ? reject(err) : resolve() )
  });
}

function all(sql) {
  return new Promise((resolve, reject) => {
      db.all(sql, (err, rows) => err ? reject(err) : resolve(rows) );
  });
}

// run our sql
async function main() {
  // create the tables and seed them with data
  console.log("\nCREATING TABLES & SEEDING WITH DATA ...");
  let files = ["test_drop", "user", "post", "comment", "test_insert"]
  for (let i = 0; i < files.length; i++) {
    await exec(fs.readFileSync(`sql/${files[i]}.sql`).toString());
  }

  // read/select all the records from all these tables
  console.log("\nREADING TABLE DATA...");
  let tables = ['user', 'post', 'comment']
  for(let i = 0; i < tables.length; i++) {
    console.log(`DATA FROM TABLE: ${tables[i]}`);
    await all(`SELECT * FROM ${tables[i]};`).then(rows => {
      rows.forEach((row) => console.log(row));
    });
  }

  // demo a sql join of all tables
  let joinQuery = "SELECT user.id, user.username, post.id, post.post_text, comment.comment_text " +
    "FROM user " +
    "JOIN post ON user.id = post.user_id " +
    "JOIN comment ON comment.post_id = comment.id " +
    "WHERE user.id = (SELECT max(id) FROM user);"

  console.log("\nJOIN RESULTS...");
  await all(joinQuery).then(rows => {
    rows.forEach((row) => console.log(row));
  });

}

main();



